core/logs.go

```diff
@@ -222,3 +226,104 @@ func (ds *dockerService) removeContainerLogSymlink(containerID string) error {
 	}
 	return nil
 }
+
+// createContainerKubeLogFile creates the kube log file for docker container log.
+func (ds *dockerService) createContainerKubeLogFile(containerID string) error {
+	kubePath, dockerPath, err := ds.getContainerLogPath(containerID)
+	if err != nil {
+		return fmt.Errorf("failed to get container %q log path: %v", containerID, err)
+	}
+
+	if kubePath == "" {
+		logrus.Debugf("Container log path for Container ID %s isn't specified, will not create kubepath", containerID)
+		return nil
+	}
+
+	if dockerPath != "" {
+		// Only create the kube log file when container log path is specified and log file exists.
+		// Delete possibly existing file first
+		if err = ds.os.Remove(kubePath); err == nil {
+			logrus.Debugf("Deleted previously existing kube log file: %s", kubePath)
+		}
+
+		_, err = os.Create(kubePath)
+		if err != nil {
+			return fmt.Errorf(
+				"failed to create the kube log file %q to the container log file %q for container %q: %v",
+				kubePath,
+				dockerPath,
+				containerID,
+				err,
+			)
+
+		}
+
+		go func() {
+			// Open the kube file for output
+			kubeFile, err := os.OpenFile(kubePath, os.O_APPEND|os.O_WRONLY|os.O_CREATE, 0600)
+			if err != nil {
+				logrus.Errorf(
+					"failed to open kube log file %s: %v",
+					kubePath,
+					err,
+				)
+				panic(err)
+			}
+			defer kubeFile.Close()
+
+			// Tail the docker file for input
+			t, err := tail.TailFile(dockerPath, tail.Config{
+				Follow: true,
+				ReOpen: true})
+			if err != nil {
+				logrus.Errorf(
+					"failed to tail docker log file %s: %v",
+					dockerPath,
+					err,
+				)
+				panic(err)
+			}
+			// Watch for changes to be copied over
+			for line := range t.Lines {
+				logLine := Log{}
+				json.Unmarshal([]byte(line.Text), &logLine)
+
+				_, err := kubeFile.WriteString(logLine.CRIFormat())
+				if err != nil {
+					logrus.Errorf(
+						"failed to write to kube log file %s: %v",
+						kubePath,
+						err,
+					)
+					return
+				}
+			}
+		}()
+	} else {
+		supported, err := ds.IsCRISupportedLogDriver()
+		if err != nil {
+			logrus.Errorf("Failed to check supported logging driver for CRI: %v", err)
+			return nil
+		}
+
+		if supported {
+			logrus.Info("Cannot create kube log file because container log file doesn't exist!")
+		} else {
+			logrus.Debug("Unsupported logging driver by CRI")
+		}
+	}
+
+	return nil
+}
+
+type Log struct {
```

Comment:
Instead of creating a fully new type, what do you think about extending the JSONLog type (since this is tied to that struct not moving anyway: https://github.com/moby/moby/blob/6ce5aa1cd5a46e02ffaa3fe0c794642b35601676/daemon/logger/jsonfilelog/jsonlog/jsonlog.go#L8) and keeping the `CRIFormat()` method on the new type?

Reply:
(Also, I think this is a good opportunity to kill `IsCRISupportedLogDriver` in favor of moving the logic into this file, and perhaps supporting `local` as well with an interface to different conversion functions.)

Reply:
I think that all sounds like a good plan. The only catch I see right now is waiting on those changes to get these tests running again.

What do you think about a comment and issue for a followup PR to replace this new type with `JSONLog`? I'm also not seeing it exported in the docker API anywhere so that will need to be added.

Reply:
It's not in the API types (since it's not visible over the API), but it's in a public (non-`internal/`) package and specifically, the type is in a minimal `github.com/docker/docker/daemon/logger/jsonfilelog/jsonlog` package that has no non-std dependencies inside the larger `github.com/docker/docker/daemon/logger/jsonfilelog` package for exactly this use-case.

I think it should be trivial to extend this PR, but I'm also fine with taking it earlier to unblock tests. In that case, I have a couple objections to the current design, but I'll need a bit more time to get them written out in a coherent way.